x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/state.ts

```diff
@@ -0,0 +1,157 @@
+import { StreamActiveRecord, StreamChange, ValidationResult } from './types';
+import { streamFromDefinition } from './stream_from_definition';
+
+export class State {
+  private streamsByName: Map<string, StreamActiveRecord>;
+
+  constructor(streams: StreamActiveRecord[]) {
+    this.streamsByName = new Map();
+    streams.forEach((stream) => this.streamsByName.set(stream.definition.name, stream));
+  }
+
+  applyChanges(requestedChanges: StreamChange[]): State {
+    // What if multiple changes target the same stream?
+
+    // Optional: Expand requested changes to include automatic creation of missing streams
+    const newState = this.clone();
+
+    requestedChanges.forEach((change) => {
+      const targetStream = newState.get(change.target);
+
+      if (!targetStream) {
+        if (change.type === 'delete') {
+          throw new Error('Cannot delete non-existing stream');
+        } else {
+          const newStream = streamFromDefinition(change.request.stream);
+          newStream.update(change.request.stream, newState, this);
+          newState.set(newStream.definition.name, newStream);
+        }
+      } else {
+        if (change.type === 'delete') {
+          targetStream.markForDeletion(newState, this);
+        } else {
+          targetStream.update(change.request.stream, newState, this);
+        }
+      }
+    });
+
+    return newState;
+  }
+
+  private clone(): State {
+    const newStreams = this.all().map((stream) => stream.clone());
+    return new State(newStreams);
+  }
+
+  async validate(
+    startingState: State,
+    scopedClusterClient: IScopedClusterClient
+  ): Promise<ValidationResult> {
+    // Should I use allSettled here?
+    const validationResults = await Promise.all(
+      this.all().map((stream) => stream.validate(this, startingState, scopedClusterClient))
+    );
+
+    const isValid = validationResults.every((validationResult) => validationResult.isValid);
+    const errors = validationResults.flatMap((validationResult) => validationResult.errors);
+
+    return {
+      isValid,
+      errors,
+    };
+  }
+
+  changes() {
+    // Could there be conflicts between changes, like a group being updated with new members before that member is deleted?
+    // I guess the deletion change for that member should catch that and update the group again or validation for the group should catch it
+    // In reverse, the change for the group should check that the member was deleted and throw? And in any case the validation should catch it
+    return this.all().filter((stream) => stream.changeStatus !== 'unchanged');
+  }
+
+  async commitChanges(
+    storageClient: StreamsStorageClient,
+    logger: Logger,
+    scopedClusterClient: IScopedClusterClient,
+    isServerless: boolean
+  ) {
+    await Promise.all(
+      this.changes().map((stream) =>
+        stream.commit(storageClient, logger, scopedClusterClient, isServerless)
+      )
+    );
+  }
+
+  async attemptRollback(
```

Comment:
This obviously requires more sophistication (such as retires within the `commit` method of each stream class) but the overall plan is to "inverse" the requested changes that have already been applied and ask the stream to commit the previous state again.

It might be needed here to better understand why the desired state commit failed and perhaps errors can be fixed before attempting the rollback as a whole.